### PR TITLE
HM-4717: Restrict closing dates during migration period

### PR DESCRIPTION
**Pull request by dta-tyler**
This pull request implements restrictions on closing dates during a specified lockout period.
- A new validator for closing date has been implemented when creating and editing an opportunity of any type. The validator checks if the closing date is during the lockout period. An error message is shown when validation fails.
- A page alert has been inserted before closing date form fields with the lockout period dates.
- The closing time is displayed as 11:55pm when an opportunity closes after the lockout period when creating or editing an opportunity, as well as on the opportunity page.
- The countdown on the opportunity page uses 11:55pm as the closing time when an opportunity ends after the lockout period.
- Lockout period dates are provided by the API. If dates are not provided by the API, alerts/validators/updated closing times are disregarded.

### CODE REVIEW

**Review by dta-tom — APPROVED** [review]
Looks good

**Comment by dta-tyler** [comment]
Opportunity time shown on completion page of each opportunity type has been corrected. Variable/function names have been changed to use 'lockout' instead of 'blackout'.

File: apps/marketplace/components/BuyerATM/BuyerATMCompleted.js [hunk 1]
```diff
@@ -1,53 +1,88 @@
...
-      <a href={`${rootPath}/brief/${props.briefId}/overview/atm`} className="au-btn au-btn--secondary">
-        Return to overview
-      </a>
-      <a href={`${rootPath}/digital-marketplace/opportunities/${props.briefId}`} className="au-btn">
-        View live opportunity
-      </a>
-    </p>
-  </div>
-)
+export class BuyerATMCompleted extends Component {
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/BuyerATM/BuyerATMCompleted.js:11]: Yes- I'll make this change.

File: apps/marketplace/components/BuyerATM/BuyerATMAdditionalInformationStage.js [hunk 2]
```diff
@@ -41,14 +51,22 @@ export class BuyerATMAdditionalInformationStage extends Component {
...
   handleDateChange(date) {
     this.props.setDate(`${date.year}-${date.month}-${date.day}`)
   }
 
   render() {
-    const { model } = this.props
+    const { model, lockoutPeriod } = this.props
+    const isLockoutPeriod = lockoutPeriod.startDate && lockoutPeriod.endDate
+    let closingTime = '6pm'
+    if (isLockoutPeriod && isAfter(new Date(this.props[this.props.model].closedAt), lockoutPeriod.startDate)) {
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/BuyerATM/BuyerATMAdditionalInformationStage.js:66]: Good pickup- will fix this.

File: apps/marketplace/components/Brief/Edit/EditOpportunityTable.js [hunk 3]
```diff
@@ -57,9 +59,25 @@ class EditOpportunityTable extends Component {
...
   render = () => {
-    const { brief, edits } = this.props
+    const { brief, edits, lockoutPeriod } = this.props
     const showInvited = this.showInvitedSellers()
     const sellersToInvite = getSellersToInvite(brief, edits)
+    let closingTime = '6pm'
+    let closingDateString = null
+
+    if (lockoutPeriod.startDate && lockoutPeriod.endDate) {
+      if (isAfter(edits.closingDate, lockoutPeriod.startDate)) {
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/Brief/Edit/EditOpportunityTable.js:69]: Will incorporate it into the same line.

File: apps/marketplace/components/Brief/Edit/EditOpportunityClosingDate.js [hunk 4]
```diff
@@ -127,9 +170,37 @@ class EditOpportunityClosingDate extends Component {
...
 
   render = () => {
-    const { brief, model } = this.props
+    const { brief, model, lockoutPeriod } = this.props
     const { hasErrors, redirectToEditsTable } = this.state
-    const invalidClosingDateMessage = <ClosingDateIsNotValidMessage closingDate={getClosingTime(brief)} />
+    const invalidClosingDateMessage = (
+      <ClosingDateIsNotValidMessage closingDate={getClosingTime(brief)} lockoutPeriod={lockoutPeriod} />
+    )
+    const isLockoutPeriod = lockoutPeriod.startDate && lockoutPeriod.endDate
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/Brief/Edit/EditOpportunityClosingDate.js:178]: I'll change the variable name to make it clearer.

File: apps/marketplace/components/Brief/Edit/EditOpportunityClosingDate.js [hunk 5]
```diff
@@ -127,9 +170,37 @@ class EditOpportunityClosingDate extends Component {
...
+        minValidDate = lockoutPeriod.endDate
+        isAfterLockoutPeriod = false
+      } else if (isAfter(closingDate, lockoutPeriod.startDate) && isBefore(closingDate, lockoutPeriod.endDate)) {
+        minValidDate = lockoutPeriod.endDate
+        showLockoutPeriod = true
+        isAfterLockoutPeriod = false
+      } else if (isBefore(closingDate, lockoutPeriod.startDate)) {
+        showLockoutPeriod = true
+        isAfterLockoutPeriod = false
+      }
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/Brief/Edit/EditOpportunityClosingDate.js:202]: I will refactor the code to reduce the significant amount of repetition.

File: apps/marketplace/components/BuyerSpecialist/BuyerSpecialistAdditionalInformationStage.js [hunk 6]
```diff
@@ -215,23 +250,30 @@ export class BuyerSpecialistRequirementsStage extends Component {
...
 const mapStateToProps = (state, props) => ({
-  ...formProps(state, props.model)
+  ...formProps(state, props.model),
+  lockoutPeriod: state.brief.lockoutPeriod
 })
 
 const mapDispatchToProps = (dispatch, props) => ({
-  setDate: date => dispatch(actions.change(`${props.model}.closedAt`, date))
+  setDate: date => dispatch(actions.change(`${props.model}.closedAt`, date)),
+  loadInitialData: briefId => dispatch(loadPublicBrief(briefId))
```
Comment (tobiasbrown) [inline apps/marketplace/components/BuyerSpecialist/BuyerSpecialistAdditionalInformationStage.js:276]: The other additional information stages don't appear to have this.  Is it required or can it be removed?

File: apps/marketplace/components/BuyerSpecialist/BuyerSpecialistReviewStage.js [hunk 7]
```diff
@@ -1,93 +1,113 @@
...
-            The last day you can publish answers to all sellers&apos; questions.
-            <br />
-            The last day sellers can apply.
-            <br />
-            The opportunity will close at 6pm Canberra time.
-          </div>
-        </div>
-      </div>
-      {props.stagesTodo.length > 0 ? (
+export class BuyerSpecialistReviewStage extends Component {
```
Comment (tobiasbrown) [inline apps/marketplace/components/BuyerSpecialist/BuyerSpecialistReviewStage.js:14]: Does this need to be a class?

File: apps/marketplace/components/BuyerTraining/BuyerTrainingReviewStage.js [hunk 8]
```diff
@@ -1,92 +1,115 @@
...
-            <div className="col-xs-12 col-sm-8">
-              Invited sellers can ask questions about your requirements and apply for the opportunity.
-            </div>
-          </div>
-          <div className="row">
-            <div className="col-xs-12 col-sm-4">
-              {format(getBriefLastQuestionDate(new Date(props[props.model].closedAt)), 'dddd D MMMM YYYY')}
-            </div>
-            <div className="col-xs-12 col-sm-8">The last day sellers can ask questions.</div>
+export class BuyerTrainingReviewStage extends Component {
```
Comment (tobiasbrown) [inline apps/marketplace/components/BuyerTraining/BuyerTrainingReviewStage.js:13]: Does this need to be a class?

File: apps/marketplace/components/BuyerATM/BuyerATMReviewStage.js [hunk 9]
```diff
@@ -1,88 +1,111 @@
...
-            <div className="col-xs-12 col-sm-8">
-              Invited sellers can ask questions about your requirements and apply for the opportunity.
-            </div>
-          </div>
-          <div className="row">
-            <div className="col-xs-12 col-sm-4">
-              {format(getBriefLastQuestionDate(new Date(props[props.model].closedAt)), 'dddd D MMMM YYYY')}
-            </div>
-            <div className="col-xs-12 col-sm-8">The last day sellers can ask questions.</div>
+export class BuyerATMReviewStage extends Component {
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/BuyerATM/BuyerATMReviewStage.js:13]: No, it doesn't need to be a class. I'll amend the affected files.

File: apps/marketplace/components/BuyerSpecialist/BuyerSpecialistAdditionalInformationStage.js (hunk 6 above)
Comment (dta-tyler) [reply apps/marketplace/components/BuyerSpecialist/BuyerSpecialistAdditionalInformationStage.js]: No it doesn't seem necessary. I can't find any references which seem dependent on the code, and when I removed it, functionality seems perfectly fine in my testing, without any errors displayed in the console.

**Comment by dta-tyler** [comment]
Recommended changes have been implemented.

**Comment by dta-tyler** [comment]
Lockout period notice has been inserted into opportunity review pages when closing date is after lockout period. Last date for questions is changed to working day prior to commencement of lockout period, for opportunities with a closing date within 2 days post-lockout.

**Content for notice still to be reviewed, and calculation of last date for questions not finalised**

File: apps/marketplace/components/BuyerTraining/BuyerTrainingReviewStage.js [hunk 10]
```diff
@@ -5,88 +5,112 @@ import { Link } from 'react-router-dom'
...
+  }
 }
 
 BuyerTrainingReviewStage.propTypes = {
   model: PropTypes.string.isRequired,
   formButtons: PropTypes.node.isRequired,
   stagesTodo: PropTypes.array,
-  onSubmit: PropTypes.func
+  onSubmit: PropTypes.func,
+  lockoutPeriod: PropTypes.objects
```
Comment (tobiasbrown) [inline apps/marketplace/components/BuyerTraining/BuyerTrainingReviewStage.js:108]:
```suggestion
  lockoutPeriod: PropTypes.object
```

Comment (dta-tyler) [reply apps/marketplace/components/BuyerTraining/BuyerTrainingReviewStage.js]: Thanks, this typo has been fixed.

**Comment by dta-tyler** [comment]
Content has been finalised and ready for final review.

**Comment by dta-tyler** [comment]
"Hard" lockout (i.e., period when users can't login):
- starts 1 day after lockout period start end
- finishes 1 day before lockout period end date

File: apps/marketplace/components/BuyerBriefFlow/LockoutReviewNotice.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        <div className="col-xs-12 col-sm-4">
+          <strong>
+            {format(startDate, 'dddd D MMMM YYYY')} – <br />
+            {format(addDays(endDate, -1), 'dddd D MMMM YYYY')}
+          </strong>
+        </div>
+        <div className="col-xs-12 col-sm-8">
+          <AUheadings level="2" size="sm">
+            Digital Marketplace is unavailable during its move to{' '}
+            <a href="/api/2/r/dmpbuyict" target="_blank" rel="noopener noreferrer">
```
[the remark this answers is not in the document]
Comment (dta-tyler) [reply apps/marketplace/components/BuyerBriefFlow/LockoutReviewNotice.js:22]: I wasn't sure if this link might need to target a different article, but it appears not, so I will change it to match the others.

**Comment by dta-tyler** [comment]
Lockout period data now passed to opportunities page, to allow countdowns for each opportunity to be correctly calculated.

**Pull request merged**